File: binson.erl
-module(binson).
-export([encode/1, decode/1]).

-define(BEGIN,      16#40). %%64
-define(END,        16#41). %%65
-define(BEGIN_ARR,  16#42). %%66
-define(END_ARR,    16#43). %%67
-define(TRUE,       16#44). %%68
-define(FALSE,      16#45). %%69
-define(DOUBLE,     16#46). %%70
-define(INT8,       16#10). %%16
-define(INT16,      16#11). %%17
-define(INT32,      16#12). %%18
-define(INT64,      16#13). %%19
-define(STR_LEN8,   16#14). %%20
-define(STR_LEN16,  16#15). %%21
-define(STR_LEN32,  16#16). %%22
-define(BYTE_LEN8,  16#18). %%24
-define(BYTE_LEN16, 16#19). %%25
-define(BYTE_LEN32, 16#1a). %%26

-define(BOUND8,  (1 bsl 7)).
-define(BOUND16, (1 bsl 15)).
-define(BOUND32, (1 bsl 31)).

%%------ Primitives encoding ------
encode(true)  -> <<?TRUE/integer>>;
encode(false) -> <<?FALSE/integer>>;

encode(Int) when is_integer(Int) ->
    if 
        Int >= -?BOUND8  andalso Int < ?BOUND8  -> <<?INT8/integer,  Int:8/integer-little>>;
        Int >= -?BOUND16 andalso Int < ?BOUND16 -> <<?INT16/integer, Int:16/integer-little>>;
        Int >= -?BOUND32 andalso Int < ?BOUND32 -> <<?INT32/integer, Int:32/integer-little>>;
        true                                    -> <<?INT64/integer, Int:64/integer-little>>
    end;

encode(Float) when is_float(Float) -> <<?DOUBLE/integer, Float/float-little>>;

encode(String) when is_list(String) -> 
    Len = length(String),
    Binary = list_to_binary(String), 
    if
        Len < 16#100   -> <<?STR_LEN8/integer,  Len:8/integer-little,  Binary/binary>>;
        Len < 16#10000 -> <<?STR_LEN16/integer, Len:16/integer-little, Binary/binary>>;
        true           -> <<?STR_LEN32/integer, Len:32/integer-little, Binary/binary>>
    end;

encode(Bytes) when is_binary(Bytes) -> 
    Len = byte_size(Bytes),
    if
        Len < 16#100   -> <<?BYTE_LEN8/integer,  Len:8/integer-little,  Bytes/binary>>;
        Len < 16#10000 -> <<?BYTE_LEN16/integer, Len:16/integer-little, Bytes/binary>>;
        true           -> <<?BYTE_LEN32/integer, Len:32/integer-little, Bytes/binary>>
    end;
     
%%------ Composites encoding ------
encode(Array) when is_tuple(Array)-> 
    Array_list = tuple_to_list(Array),
    Array_data = encode_array(Array_list, <<>>),
    <<?BEGIN_ARR/integer, Array_data/binary, ?END_ARR/integer>>;

encode(Object) when is_map(Object)-> 
    Fields = lists:sort(maps:keys(Object)),
    Obj_data = encode_object(Fields, Object, <<>>),
    <<?BEGIN/integer, Obj_data/binary, ?END/integer>>.

%%------ Composites encoding additional functions ------
encode_object([], _Object, Acc) -> Acc;
encode_object([Key|Tail], Object, Acc) ->
    Value = maps:get(Key, Object),
    Key_data = encode(Key),
    Value_data = encode(Value),
    encode_object(Tail, Object, <<Acc/binary, Key_data/binary, Value_data/binary>>).

encode_array([], Acc) -> Acc;
encode_array([Value|Tail], Acc) ->
    Value_data = encode(Value),
    encode_array(Tail, <<Acc/binary, Value_data/binary>>).

%%------ Primitives decoding ------

decode(<<?TRUE,  Rest/binary>>) -> {true, Rest};
decode(<<?FALSE, Rest/binary>>) -> {false, Rest};

decode(<<?INT8,  Int:8/integer-little,  Rest/binary>>) -> {Int, Rest};
decode(<<?INT16, Int:16/integer-little, Rest/binary>>) -> {Int, Rest};
decode(<<?INT32, Int:32/integer-little, Rest/binary>>) -> {Int, Rest};
decode(<<?INT64, Int:64/integer-little, Rest/binary>>) -> {Int, Rest};

decode(<<?DOUBLE, Float/float-little, Rest/binary>>) -> {Float, Rest};

decode(<<?STR_LEN8,  Len:8/integer-little,  Str:Len/binary, Rest/binary>>) -> {binary_to_list(Str), Rest};
decode(<<?STR_LEN16, Len:16/integer-little, Str:Len/binary, Rest/binary>>) -> {binary_to_list(Str), Rest};
decode(<<?STR_LEN32, Len:32/integer-little, Str:Len/binary, Rest/binary>>) -> {binary_to_list(Str), Rest};

decode(<<?BYTE_LEN8,  Len:8/integer-little,  Bytes:Len/binary, Rest/binary>>) -> {Bytes, Rest};
decode(<<?BYTE_LEN16, Len:16/integer-little, Bytes:Len/binary, Rest/binary>>) -> {Bytes, Rest};
decode(<<?BYTE_LEN32, Len:32/integer-little, Bytes:Len/binary, Rest/binary>>) -> {Bytes, Rest};

%%------ Composites decoding ------
decode(<<?BEGIN_ARR, Rest/binary>>) -> decode_array(Rest, []);
decode(<<?BEGIN, Rest/binary>>) -> decode_object(Rest, #{}).

%%------ Composites decoding additional functions ------
decode_object(<<?END, Rest/binary>>, Acc) -> {Acc, Rest};
decode_object(Data, Acc) ->
    {Key, Val_Rest} = decode(Data),
    {Value, Rest}   = decode(Val_Rest),
    decode_object(Rest, maps:put(Key, Value, Acc)).

decode_array(<<?END_ARR, Rest/binary>>, Acc) -> {list_to_tuple(lists:reverse(Acc)), Rest};
decode_array(Data, Acc) ->
    {Value, Rest}   = decode(Data),
    decode_array(Rest, [Value|Acc]).

File: binson.ex
defmodule Binson do
    @bs_begin       0x40  #64
    @bs_end         0x41  #65
    @bs_begin_arr   0x42  #66
    @bs_end_arr     0x43  #67
    @bs_true        0x44  #68
    @bs_false       0x45  #69
    @bs_double      0x46  #70
    @bs_int8        0x10  #16
    @bs_int16       0x11  #17
    @bs_int32       0x12  #18
    @bs_int64       0x13  #19
    @bs_str_len8    0x14  #20
    @bs_str_len16   0x15  #21
    @bs_str_len32   0x16  #22
    @bs_byte_len8   0x18  #24
    @bs_byte_len16  0x19  #25
    @bs_byte_len32  0x1a  #26

    @bs_bound8     0x80         # 1<<7
    @bs_bound16    0x8000       # 1<<15
    @bs_bound32    0x80000000   # 1<<31

#------ Primitives encoding ------
    def encode(true),  do: <<@bs_true>>
    def encode(false), do: <<@bs_false>>

    def encode(int) when is_integer(int) do
        cond do
            (int >= -@bs_bound8)  and (int < @bs_bound8)  -> <<@bs_int8,  int::size(8)-signed-integer-little>>
            (int >= -@bs_bound16) and (int < @bs_bound16) -> <<@bs_int16, int::size(16)-signed-integer-little>>
            (int >= -@bs_bound32) and (int < @bs_bound32) -> <<@bs_int32, int::size(32)-signed-integer-little>>
            true                                          -> <<@bs_int64, int::size(64)-signed-integer-little>>
        end
    end

    def encode(float) when is_float(float), do: <<@bs_double, float::float-little>>;

    def encode(string) when is_binary(string) do
        len = byte_size(string)
        cond do
            len < @bs_bound8  -> <<@bs_str_len8,  len::size(8)-integer-little,  string::binary>>
            len < @bs_bound16 -> <<@bs_str_len16, len::size(16)-integer-little, string::binary>>
            true              -> <<@bs_str_len32, len::size(32)-integer-little, string::binary>>
        end
    end

    def encode({:bin, bytes}) do
        len = byte_size(bytes)
        cond do
            len < @bs_bound8  -> <<@bs_byte_len8,  len::size(8)-integer-little,  bytes::binary>>
            len < @bs_bound16 -> <<@bs_byte_len16, len::size(16)-integer-little, bytes::binary>>
            true              -> <<@bs_byte_len32, len::size(32)-integer-little, bytes::binary>>
        end
    end

#------ Composites encoding ------
    def encode(array) when is_list(array), do: encode_array(array, <<@bs_begin_arr>>)

    def encode(object) when is_map(object) do
        Map.to_list(object) |> List.keysort(0) |> encode_object(<<@bs_begin>>)
    end

#------ Composites encoding private functions ------
    defp encode_array([], acc), do: <<acc::binary, @bs_end_arr>>
    defp encode_array([value|tail], acc) do
        encode_array(tail, <<acc::binary, encode(value)::binary>>)
    end

    defp encode_object([], acc), do: <<acc::binary, @bs_end>>
    defp encode_object([{key, value}|tail], acc) do
        encode_object(tail, <<acc::binary, encode(key)::binary, encode(value)::binary>>)
    end

#------ Primitives decoding ------
    def decode(<<@bs_true,  rest::binary>>) , do: {true, rest}
    def decode(<<@bs_false, rest::binary>>) , do: {false, rest}

    def decode(<<@bs_int8,  int::size(8)-signed-integer-little,  rest::binary>>), do: {int, rest}
    def decode(<<@bs_int16, int::size(16)-signed-integer-little, rest::binary>>), do: {int, rest}
    def decode(<<@bs_int32, int::size(32)-signed-integer-little, rest::binary>>), do: {int, rest}
    def decode(<<@bs_int64, int::size(64)-signed-integer-little, rest::binary>>), do: {int, rest}

    def decode(<<@bs_double, float::float-little, rest::binary>>), do: {float, rest}

    def decode(<<@bs_str_len8,  len::size(8)-integer-little,  str::size(len)-binary, rest::binary>>), do: {str, rest}
    def decode(<<@bs_str_len16, len::size(16)-integer-little, str::size(len)-binary, rest::binary>>), do: {str, rest}
    def decode(<<@bs_str_len32, len::size(32)-integer-little, str::size(len)-binary, rest::binary>>), do: {str, rest}

    def decode(<<@bs_byte_len8,  len::size(8)-integer-little,  bytes::size(len)-binary, rest::binary>>), do: {{:bin, bytes}, rest}
    def decode(<<@bs_byte_len16, len::size(16)-integer-little, bytes::size(len)-binary, rest::binary>>), do: {{:bin, bytes}, rest}
    def decode(<<@bs_byte_len32, len::size(32)-integer-little, bytes::size(len)-binary, rest::binary>>), do: {{:bin, bytes}, rest}

#------ Composites decoding ------
    def decode(<<@bs_begin_arr, rest::binary>>), do: decode_array(rest, [])
    def decode(<<@bs_begin,     rest::binary>>), do: decode_object(rest, %{})

#------ Composites decoding private functions ------
    defp decode_object(<<@bs_end, rest::binary>>, acc), do: {acc, rest}
    defp decode_object(data, acc) do
        {key, val_rest} = decode(data)
        {value, rest}   = decode(val_rest)
        decode_object(rest, Map.put(acc, key, value))
    end

    defp decode_array(<<@bs_end_arr, rest::binary>>, acc), do: {:lists.reverse(acc), rest}
    defp decode_array(data, acc) do
        {value, rest} = decode(data)
        decode_array(rest, [value|acc])
    end
end
